src/proxy/http/metrics/handle_time.rs

```diff
@@ -0,0 +1,252 @@
+use metrics::{latency, FmtLabels, FmtMetric, Histogram};
+use proxy::http::insert;
+use std::{
+    fmt,
+    sync::{
+        atomic::{self, AtomicUsize, Ordering},
+        Arc, Mutex, RwLock,
+    },
+    time::Instant,
+};
+
+/// A single handle time histogram.
+///
+/// Higher-level code will use this to represent a single set of labels for
+/// handle-time metrics.
+#[derive(Debug, Clone)]
+pub struct Scope(Arc<Shared>);
+
+/// A layer that inserts a `Tracker` into each request passing through it.
+#[derive(Debug, Clone)]
+pub struct InsertTracker(Arc<Shared>);
+
+/// A request extension that, when dropped, records the time elapsed since it
+/// was created.
+#[derive(Debug)]
+pub struct Tracker {
+    shared: Arc<Shared>,
+    idx: usize,
+    t0: Instant,
+}
+
+#[derive(Debug)]
+struct Shared {
+    // NOTE: this is inside a `Mutex` since recording a latency requires a mutable
+    // reference to the histogram. In the future, we could consider making the
+    // histogram counters `AtomicU64, so that the histogram could be updated
+    // with an immutable reference. Then, the mutex could be removed.
+    histogram: Mutex<Histogram<latency::Us>>,
+    /// Stores the state of currently active `Tracker`s.
+    recorders: RwLock<Vec<Recorder>>,
+    /// The index of the most recently finished recorder.
+    ///
+    /// When a new recorder is needed, the recorder at this index will be used,
+    /// and that recorder's next index will be set as the head. of the free
+    /// list.
+    ///
+    /// When an active recorder completes, this will be set to its index, and
+    /// the previous value will become the freed recorder's next pointer.
+    idle_head: AtomicUsize,
+}
+
+#[derive(Debug)]
+struct Recorder {
```

Comment:
I don't really understand the role of `Recorder` in this. My naive assumption is that this should all work without needing to store the rwlock of recorders. When would an individual Recorder have more than one tracker? Why would a scope have more than one recorder?

It would be helpful if the comments addressed this more directly.

Reply:
Requests may be cloned for retries, and we want to include any retries in the handle time for a request. The scope will have one `Recorder` for each request that is currently being tracked, and if that request is cloned, there will be a `Tracker` inserted into each clone's extension, incrementing the count. I changed some naming & added comments in 159b6730b9d928cb213cf4ee2b1f00657f385f76, hopefully the purpose of this is a little clearer.

The alternative to using the slab of counters is to create a new `Arc<()>` for each tracker, and record the handle time when that `Arc` is dropped; this would mean making a small heap allocation for every request. The slab lets us amortize the cost of storing the counters.